File: packages/ws-proto-transport/test/recvbound.test.ts
import { describe, it, expect } from "vitest";
import { WsTransport, FakeSocket, Kind } from "../src/index.js";

/** tick yields to the microtask/timer queue so async plumbing settles. */
const tick = () => new Promise<void>((r) => setTimeout(r, 0));

describe("byte-based receive bound", () => {
  it("does not reset a window-obeying peer that sends many small messages", async () => {
    const sock = new FakeSocket();
    // Default 256 KiB window. 300 one-byte messages = 300 bytes, far under it —
    // a window-obeying peer would send all of these without waiting for credit.
    const t = WsTransport.fromSocket(sock);
    const s = t.openStream("/s/Many");
    await tick();

    const n = 300;
    for (let i = 0; i < n; i++) {
      sock.inject({ streamId: s.id, kind: Kind.KIND_MSG, payload: new Uint8Array([i & 0xff]) });
    }
    sock.inject({ streamId: s.id, kind: Kind.KIND_END, status: { code: 0 } });

    let count = 0;
    for (;;) {
      const m = await s.recv();
      if (m === null) {
        break;
      }
      count++;
    }
    expect(count).toBe(n);

    // The stream must NOT have been reset: more frames than a 256-FRAME cap, but
    // the byte-based bound (aligned with the window) is nowhere near exceeded.
    const rst = sock.sent.find((f) => f.kind === Kind.KIND_RST && f.streamId === s.id);
    expect(rst).toBeUndefined();
  });
});

File: wsrpc/recvbound_test.go
package wsrpc

import (
	"context"
	"io"
	"testing"
	"time"

	"github.com/gopherex/ws-proto/transport"
	"github.com/stretchr/testify/require"
	"google.golang.org/protobuf/types/known/wrapperspb"
)

// TestReceiveBoundIsByteBasedNotFrameCount reproduces M3: the receive backlog
// must be bounded by BYTES (aligned with the flow-control window), not by a
// fixed frame count. With the default 256 KiB window a well-behaved peer may
// have thousands of tiny messages in flight; a 256-FRAME cap would falsely reset
// such a stream with ResourceExhausted even though it obeys the window.
func TestReceiveBoundIsByteBasedNotFrameCount(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	gotStream := make(chan *Stream, 1)
	srvEnd, peer := newPipe()
	_ = newMuxBuffered(ctx, srvEnd, func(s *Stream) { gotStream <- s }, defaultReceiveBuffer)

	s := openServerStream(t, ctx, peer, gotStream, 1, "/t/M")

	// 300 tiny messages: far more than the 256-frame count, but only ~900 bytes
	// total — well under the 256 KiB window. A window-obeying peer would send
	// all of these without waiting for credit.
	const n = 300
	for i := 0; i < n; i++ {
		require.NoError(t, peer.WriteFrame(ctx, &transport.Frame{
			StreamId: 1, Kind: transport.Kind_KIND_MSG, Payload: mustMarshalString(t, "x"),
		}))
	}
	require.NoError(t, peer.WriteFrame(ctx, &transport.Frame{StreamId: 1, Kind: transport.Kind_KIND_HALF_CLOSE}))

	rctx, rcancel := context.WithTimeout(ctx, 2*time.Second)
	defer rcancel()
	count := 0
	for {
		var v wrapperspb.StringValue
		err := s.Recv(&v)
		if err == io.EOF {
			break
		}
		require.NoError(t, err, "window-obeying peer was falsely reset after %d msgs", count)
		count++
		_ = rctx
	}
	require.Equal(t, n, count)
}
